src/adc.rs

```diff
@@ -0,0 +1,407 @@
+//! # API for the Analog to Digital converter
+
+use embedded_hal::adc::{Channel, OneShot};
+
+use crate::gpio::Analog;
+use crate::gpio::{gpioa, gpiob, gpioc};
+use crate::rcc::APB2;
+
+use crate::stm32::ADC1;
+#[cfg(any(
+    feature = "stm32f103",
+))]
+use crate::stm32::ADC2;
+
+/// ADC configuration
+pub struct Adc<ADC> {
+    rb: ADC,
+    sample_time: AdcSampleTime,
+    align: AdcAlign,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+/// ADC sampling time
+///
+/// Options for the sampling time, each is T + 0.5 ADC clock cycles.
+pub enum AdcSampleTime {
+    /// 1.5 cycles sampling time
+    T_1,
+    /// 7.5 cycles sampling time
+    T_7,
+    /// 13.5 cycles sampling time
+    T_13,
+    /// 28.5 cycles sampling time
+    T_28,
+    /// 41.5 cycles sampling time
+    T_41,
+    /// 55.5 cycles sampling time
+    T_55,
+    /// 71.5 cycles sampling time
+    T_71,
+    /// 239.5 cycles sampling time
+    T_239,
+}
+
+impl AdcSampleTime {
+    /// Get the default sample time (currently 28.5 cycles)
+    pub fn default() -> Self {
+        AdcSampleTime::T_28
+    }
+}
+
+impl From<AdcSampleTime> for u8 {
+    fn from(val: AdcSampleTime) -> Self {
+        match val {
+            AdcSampleTime::T_1 => 0,
+            AdcSampleTime::T_7 => 1,
+            AdcSampleTime::T_13 => 2,
+            AdcSampleTime::T_28 => 3,
+            AdcSampleTime::T_41 => 4,
+            AdcSampleTime::T_55 => 5,
+            AdcSampleTime::T_71 => 6,
+            AdcSampleTime::T_239 => 7,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+/// ADC data register alignment
+pub enum AdcAlign {
+    /// Right alignment of output data
+    Right,
+    /// Left alignment of output data
+    Left,
+}
+
+impl AdcAlign {
+    /// Default: right alignment
+    pub fn default() -> Self {
+        AdcAlign::Right
+    }
+}
+
+impl From<AdcAlign> for u8 {
+    fn from(val: AdcAlign) -> Self {
+        match val {
+            AdcAlign::Right => 0,
+            AdcAlign::Left => 1,
+        }
+    }
+}
+
+impl From<AdcAlign> for bool {
+    fn from(val: AdcAlign) -> Self {
+        match val {
+            AdcAlign::Right => false,
+            AdcAlign::Left => true,
+        }
+    }
+}
+
+macro_rules! adc_pins {
+    ($ADC:ident, $($pin:ty => $chan:expr),+ $(,)*) => {
+        $(
+            impl Channel<$ADC> for $pin {
+                type ID = u8;
+
+                fn channel() -> u8 { $chan }
+            }
+        )+
+    };
+}
+
+#[cfg(any(
```

Comment:
Since ADC1 is available on all pins, perhaps we can skip this

Reply:
Ok. Will do.